src/block.rs

```diff
@@ -47,23 +93,81 @@ pub struct BlockHeader {
     /// Block height
     pub height: u32,
     /// Block height
-    pub proof: Proof,
+    pub ext: ExtData,
+
+}
+
+impl<S: consensus::Encoder> consensus::Encodable<S> for BlockHeader {
+    fn consensus_encode(&self, s: &mut S) -> Result<(), consensus::encode::Error> {
+        let version = if let ExtData::Dynafed { .. } = self.ext {
+            self.version | 0x8000_0000
+        } else {
+            self.version
+        };
+        version.consensus_encode(s)?;
+        self.prev_blockhash.consensus_encode(s)?;
+        self.merkle_root.consensus_encode(s)?;
+        self.time.consensus_encode(s)?;
+        self.height.consensus_encode(s)?;
+        self.ext.consensus_encode(s)?;
+        Ok(())
+    }
+}
+
+impl<D: consensus::Decoder> consensus::Decodable<D> for BlockHeader {
+    fn consensus_decode(d: &mut D) -> Result<Self, consensus::encode::Error> {
+        let mut version: u32 = consensus::Decodable::consensus_decode(d)?;
+        let is_dyna = if version >> 31 == 1 {
+            version &= 0x7fff_ffff;
+            true
+        } else {
+            false
+        };
 
+        Ok(BlockHeader {
+            version: version,
+            prev_blockhash: consensus::Decodable::consensus_decode(d)?,
+            merkle_root: consensus::Decodable::consensus_decode(d)?,
+            time: consensus::Decodable::consensus_decode(d)?,
+            height: consensus::Decodable::consensus_decode(d)?,
+            ext: if is_dyna {
+                ExtData::Dynafed {
+                    current: consensus::Decodable::consensus_decode(d)?,
+                    proposed: consensus::Decodable::consensus_decode(d)?,
+                    signblock_witness: consensus::Decodable::consensus_decode(d)?,
+                }
+            } else {
+                ExtData::BlockProof(consensus::Decodable::consensus_decode(d)?)
+            },
+        })
+    }
 }
-impl_consensus_encoding!(BlockHeader, version, prev_blockhash, merkle_root, time, height, proof);
 
 impl BitcoinHash for BlockHeader {
     fn bitcoin_hash(&self) -> sha256d::Hash {
         use bitcoin::consensus::Encodable;
 
+        let version = if let ExtData::Dynafed { .. } = self.ext {
+            self.version | 0x8000_0000
+        } else {
+            self.version
+        };
+
         // Everything except proof.solution goes into the hash
```

Comment:
Maybe update this to "everything except the signblock witness goes into the hash".